### ISSUE TITLE
XMLHttpRequest Events in lib.dom don't match spec[?

### ISSUE THREAD

**Issue by ericdrobinson** [Microsoft/TypeScript#21414]
**TypeScript Version:**  2.7.0-dev.201xxxxx

**Search Terms:** 

**Code**

```ts
// A *self-contained* demonstration of the problem follows...
// Test this by running `tsc` on the command-line, rather than through another build tool such as Gulp, Webpack, etc.
```

**Expected behavior:**

**Actual behavior:**

**Playground Link:**  

**Related Issues:**

**Comment by ericdrobinson** [comment]
@mhegazy Got it. Thanks.

Followup question(s):

1. How are those spec files produced? Is there a command in Edge to make them?
1. Where is the WebIDL Schema to which those files adhere? Looking at the links in the root element results in the following unhelpful error:
  ```The resource you are looking for has been removed, had its name changed, or is temporarily unavailable.```

**Comment by mhegazy** [comment]
> How are those spec files produced?

we grab them from the Edge build output. it is internal at the moment, but we have plans to move to standard webidl in the future.

> Where is the WebIDL Schema

again, Edge internal docs, we do not have access to, nor do we (as the TS team) own.

**Comment by alvincrespo** [comment]
I'm curious if [this](https://www.typescriptlang.org/play/#src=const%20xhr%20%3D%20new%20XMLHttpRequest()%3B%0D%0A%0D%0Axhr.onload%20%3D%20function%20(event%3A%20ProgressEvent)%20%7B%0D%0A%20%20%20%20console.dir(event.target.status)%3B%0D%0A%20%20%20%20console.dir(event.target.response)%3B%0D%0A%7D%0D%0A) issue is related.

Basically `status` and `response` are not being picked up by `target`, which should be of `XMLHTTPRequest`. 

Should I open up a new issue?

**Comment by ericdrobinson** [comment]
@alvincrespo No, that's a different issue. The [`ProgressEvent` type](https://developer.mozilla.org/en-US/docs/Web/API/ProgressEvent) is not specific to `XMLHTTPRequest`. TypeScript could _potentially_ figure that out by analyzing the _entire_ line and noticing that `xhr.onload` is a callback for an `XMLHTTPRequest` object and therefore the `event.target` in the callback _should_ be an `XMLHTTPRequest` instance, but that seems like a pretty large bit of work (perhaps @mhegazy could say better whether that type of inference is feasible in TypeScript).

The proper way to handle the situation you pointed out can be seen [here](https://www.typescriptlang.org/play/#src=const%20xhr%20%3D%20new%20XMLHttpRequest()%3B%0D%0A%0D%0Axhr.onload%20%3D%20function%20(event%3A%20ProgressEvent)%20%7B%0D%0A%20%20%20%20console.dir(event.target.status)%3B%0D%0A%20%20%20%20console.dir(event.target.response)%3B%0D%0A%7D%0D%0A). Another example that relies upon closures would look like [this](https://www.typescriptlang.org/play/#src=const%20xhr%20%3D%20new%20XMLHttpRequest()%3B%0D%0A%0D%0Axhr.onload%20%3D%20function%20(event%3A%20ProgressEvent)%20%7B%0D%0A%20%20%20%20console.dir(event.target.status)%3B%0D%0A%20%20%20%20console.dir(event.target.response)%3B%0D%0A%7D%0D%0A). (Personally I'm partial to the non-closure approach...)

Hope this helps!

**Comment by alvincrespo** [comment]
@ericdrobinson The two links you sent over still link to my example, perhaps a copy/paste typo?

Alos, the explanation you gave makes total sense - thanks for pointing that out. More than happy to help with this problem.

**Comment by ericdrobinson** [comment]
@alvincrespo Sorry about that! I forgot to click the "Share" button to update the link before copying the link in the address bar (both times!). The links in the original comment are fixed!

**Comment by alvincrespo** [comment]
@ericdrobinson Oh cool! I totally didn't think about the casting approach! Thanks for the tip. Also, sorry for derailing the original conversation in this thread (to anyone who comes back to read this).
